plugins/auto-sizes/hooks.php

```diff
@@ -64,28 +64,46 @@ function auto_sizes_update_content_img_tag( $html ): string {
 		return $html;
 	}
 
-	// Bail early if the image is not lazy-loaded.
-	$value = $processor->get_attribute( 'loading' );
-	if ( ! is_string( $value ) || 'lazy' !== strtolower( trim( $value, " \t\f\r\n" ) ) ) {
-		return $html;
+	auto_sizes_apply_tag_parser_updates( $processor );
```

Comment:
The reason for passing around the processor is that there are two different code paths to this logic being needed:

* When updating an IMG tag in isolation via the `wp_content_img_tag` filter.
* When updating an IMG tag in the context of an entire document that is currently being iterated over in Optimization Detective. 

In Optimization Detective the tag processor is passed to each of the registered tag visitors for each open tag while iterating over the document. See the `auto_sizes_visit_tag()` function in the diff below for how it is being used there.

> There's the danger here of making code appear to change without explanation, because something the process was sent to leads to its internal state changing, and those transitions aren't evident here where the processor was created.

I don't understand. What's the danger?

Reply:
this is, by the way, a problem that eventually we'll want to solve in Core. I have no idea what the answer might look like today, but it's in the realm of "limited access processors" that could be handed off in a WordPress filter, for example.

here's an example of where things can go fishy.

```php
$processor = new WP_HTML_Tag_Processor( $html );
$total_images = 0;
while ( $processor->next_tag( 'IMG' ) ) {
	++$total_images;
	do_something( $processor );
}
```

At the end of this script we only see `$total_images === 1` and wonder why it's so wrong, given that we sent 1,000 images to it. Well, `do_something()` called `$processor->next_tag( 'SOURCE' )` but there weren't any, so by the time it returns control to the calling loop, the Tag Processor has reached the end of the document and skips all remaining images.

Reply:
Thanks for that. Optimization Detective is actually accounting for that by setting a bookmark on the tag prior to invoking all the tag visitors, and then after they have been invoked it checks if the cursor has moved at all, and if so, it seems back to the bookmark before continuing to the next tag:

https://github.com/WordPress/performance/blob/4d2299803fb5ac397ca0df14567d7b4d5791caad/plugins/optimization-detective/optimization.php#L217-L232

The processor subclass introduces `$processor->get_seek_count()` and `$processor->get_next_token_count()` (the name of the latter I see is confusing) which return the number of times 'seek()` and 'next_token()` were called.

Reply:
this is all baked into "I trust you" but in general, I like noting it, and don't think it's a terrible idea to caution others against following the example. it looks deceivingly simpler than it is.